cmd/harmony/config_migrations.go

```diff
@@ -0,0 +1,81 @@
+package main
+
+import (
+	"errors"
+
+	"github.com/pelletier/go-toml"
+)
+
+func migrateConf(confBytes []byte) (harmonyConfig, string, error) {
+	confTree, err := toml.LoadBytes(confBytes)
+	if err != nil {
+		return harmonyConfig{}, "", err
+	}
+
+	confVersion := confTree.Get("Version").(string)
+	if confVersion == "" {
+		return harmonyConfig{}, "", errors.New("invalid config file - version missing")
+	}
+	migratedVer := confVersion
```

Comment:
> Thanks for your effort in this bounty.
> 
> #### For your questions:
> > one key factor that might be worth discussing is what to be done with file versions that are not in the migrations map. Currently all versions that are not found are being handled with same migration as of 1.0.4. Some of the other options would be:
> > 
> > * adding migrations for previous versions since there probably are not that many previous versions
> > * comparing the with current and older versions are migrated from the first migration available while for newer versions (probably old version bin with new config) either error is returned or warning and loading the default.
> 
> Adding migrations between v1.0.0 to v1.0.4 is optional - they can be treated as v1.0.4 for all the hackings to make them backward compatible. But it would be definitely better to include previous version of config in the migration unit tests.

Ok, so I will build some old versions and dump the config and add it to the test cases

> 
> > Another is whether flags should match exactly config like dnssync.port or should stay as they are dns.port . Currently they are as they were ex - dns.port
> 
> `--dns.port` is fine to be left as it is. But some other flags would be better to be renames, E.g. `sync.server` -> `dns.server`, e.t.c. Please also note that in the PR just merged (#3669), `--dns.port` was changed default value from 9000 to 9000-3000=6000. Please be notified of the change
> 
> For newer versions, it would be definitely better to warn user about the version issue, and then load the config as latest in binary. So would you please introduce [go-version](https://github.com/hashicorp/go-version) for version string validation and comparison?

Ok, will use go-version to make sure that it is a valid semantic version. Older then 1.0.4 will be handled as 1.0.4 while any unknown version will print warning and load the defaults

> 
> #### For PR reviews: generally looks good to me. Some comments:
> 1. Please fix the build errors to have this PR passed in travis CICD test.
> 2. Please rebase this branch on main after merging #3669. In this PR, v1.0.5 is introduced to update port number in `network.DNSPort` (9000 -> 6000), and added `sync.legacyServer.Port` (defValue: 6000). I am sorry for not including this PR in the bounty description in the first place. BTW, since v1.0.5 is not released, so you are free to integrate it along with v2.0.0 update.

Just to clarify this - any config prior to 2.0.0 containing network.DNSPort == 9000 should be changed to 6000 ? If so that can also be part of the migration

> 3. For flags, it's recommended to update the fields if it makes sense. For example, `--dns.port` -> `--dns.zone.port`, `--sync.legacy.server` -> `--dns.server`, e.t.c. It is also recommended to have deprecated flags for maximum backward compatibility.
> 4. Do you mind dump the previous version toml files to integrate them with unit tests?

Ok I will rebase and resolve the issues that has been discussed so far